File: apps/flutter-app/lib/features/transactions/data/repositories/transaction_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:fondos_btg/core/error/exceptions.dart';
import 'package:fondos_btg/core/error/failures.dart';
import 'package:fondos_btg/features/transactions/data/datasources/transaction_remote_datasource.dart';
import 'package:fondos_btg/features/transactions/domain/entities/transaction.dart';
import 'package:fondos_btg/features/transactions/domain/repositories/transaction_repository.dart';

class TransactionRepositoryImpl implements TransactionRepository {
  final TransactionRemoteDatasource _remoteDatasource;

  const TransactionRepositoryImpl(this._remoteDatasource);

  @override
  Future<Either<Failure, List<FundTransaction>>> getTransactions() async {
    try {
      final dtos = await _remoteDatasource.getTransactions();
      final transactions = dtos.map((dto) => dto.toEntity()).toList();
      // Sort by most recent first
      transactions.sort((a, b) => b.createdAt.compareTo(a.createdAt));
      return Right(transactions);
    } on ServerException catch (e) {
      return Left(ServerFailure(e.message));
    } on NotFoundException catch (e) {
      return Left(NotFoundFailure(e.message));
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }
}

File: apps/angular-app/src/app/features/transactions/data/repositories/transaction.repository-impl.ts
import { Injectable, inject } from '@angular/core';
import { Observable, map, catchError, of, forkJoin } from 'rxjs';
import { Result, right, left } from '../../../../core/errors/either';
import { ServerFailure } from '../../../../core/errors/failures';
import { Transaction } from '../../domain/entities/transaction.entity';
import { TransactionRepository } from '../../domain/repositories/transaction.repository';
import { TransactionRemoteDatasource } from '../datasources/transaction-remote.datasource';
import { TransactionMapper } from '../mappers/transaction.mapper';

@Injectable({ providedIn: 'root' })
export class TransactionRepositoryImpl extends TransactionRepository {
  private readonly datasource = inject(TransactionRemoteDatasource);

  getTransactions(): Observable<Result<Transaction[]>> {
    return forkJoin({
      transactions: this.datasource.getTransactions(),
      funds: this.datasource.getFunds(),
    }).pipe(
      map(({ transactions, funds }) => {
        const sorted = [...transactions].sort(
          (a, b) => new Date(b.date).getTime() - new Date(a.date).getTime()
        );
        return right<Transaction[]>(TransactionMapper.fromDtoList(sorted, funds));
      }),
      catchError((error) => {
        return of(left<ServerFailure>(new ServerFailure(error.message)));
      })
    );
  }
}
